### PR TITLE
Merge changes from a single DidChange notification

### PR DESCRIPTION
**Pull request by ToddGrun**
I'm currently investigating a customer profile where didChange notifications are showing very prevalently in CPU time. I've create this PR (https://github.com/dotnet/roslyn/pull/74267) to help with the underlying performance of creating the SourceText, however, there would still be potentially many source texts created for a single operation.

This PR attempts to alleviate that situation by changing the DidChange handler to create a single source text on the input TextDocumentContentChangeEvent array. This array can be quite large, such as when the user does a replace-all or multi-line edit in VS.

Roughly 50% of CPU time in the customer profile was spent in our DidChange handler (about 29 seconds of CPU time).

image

### CODE REVIEW

File: src/LanguageServer/Protocol/Extensions/Extensions.cs [hunk 1]
```diff
@@ -243,5 +243,20 @@ private static bool TryGetVSCompletionListSetting(ClientCapabilities clientCapab
 
             return true;
         }
+
+        public static int CompareTo(this Position p1, Position p2)
```
Comment (ToddGrun) [inline src/LanguageServer/Protocol/Extensions/Extensions.cs:247]: >CompareTo [](http://example.com/codeflow?start=26&length=9)

I'm sure there is a similar method implemented somewhere in Roslyn, but I couldn't find it

**Comment by ToddGrun** [comment]
@dotnet/roslyn-ide -- ptal

**Comment by ToddGrun** [comment]
@dotnet/roslyn-ide -- ptal

**Comment by ToddGrun** [comment]
@dibarbet or @CyrusNajmabadi -- ptal

**Comment by ToddGrun** [comment]
@dotnet/roslyn-ide -- ptal

File: src/LanguageServer/Protocol/Handler/DocumentChanges/DidChangeHandler.cs [hunk 2]
```diff
@@ -28,15 +30,52 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(DidChangeTextDocumentPar
...
+        // If the host sends us changes in a way such that no earlier change can affect the position of a later change,
+        // then we can merge the changes into a single TextChange, allowing creation of only a single new
+        // source text.
+        for (var i = 1; i < contentChanges.Length; i++)
+        {
+            var prevChange = contentChanges[i - 1];
+            var curChange = contentChanges[i];
 
-        context.UpdateTrackedDocument(request.TextDocument.Uri, text);
+            if (prevChange.Range.Start.CompareTo(curChange.Range.End) < 0)
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/Handler/DocumentChanges/DidChangeHandler.cs:57]: Its early and I'm not 100% sure of this.  Is it possible we need to compare against all previous changes?  For example we have two disparate changes, then one afterwards that overlaps just the first one?

**Review by dibarbet — COMMENTED** [review]
sorry, somehow this one slipped passed me

File: src/LanguageServer/Protocol/Handler/DocumentChanges/DidChangeHandler.cs (hunk 2 above)
Comment (ToddGrun) [reply src/LanguageServer/Protocol/Handler/DocumentChanges/DidChangeHandler.cs]: Chatted in person, let me know if you still have concerns.

File: src/LanguageServer/Protocol/Handler/DocumentChanges/DidChangeHandler.cs [hunk 3]
```diff
@@ -28,15 +30,52 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(DidChangeTextDocumentPar
...
+        // positions between changes. See
         // https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#didChangeTextDocumentParams
         // for more details.
-        foreach (var change in request.ContentChanges)
-            text = text.WithChanges(ProtocolConversions.ContentChangeEventToTextChange(change, text));
+        //
+        // If the host sends us changes in a way such that no earlier change can affect the position of a later change,
+        // then we can merge the changes into a single TextChange, allowing creation of only a single new
+        // source text.
+        for (var i = 1; i < contentChanges.Length; i++)
```
[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/LanguageServer/Protocol/Handler/DocumentChanges/DidChangeHandler.cs]: Added some unit tests, they aren't as clean as I had hoped, but it should at least give some coverage

**Comment by ToddGrun** [comment]
@dibarbet -- any remaining concerns?

**Review by dibarbet — APPROVED** [review]

**Pull request merged**